src/device-state.coffee

```diff
@@ -0,0 +1,512 @@
+		@stepsInProgress.push(step)
+		setImmediate =>
+			@executeStepAction(step, { force, targetState })
+			.finally =>
+				Promise.using @inferStepsLock(), =>
+					_.pullAllWith(@stepsInProgress, [ step ], _.isEqual)
+			.then (stepResult) =>
+				@emitAsync('step-completed', null, step, stepResult)
+				@continueApplyTarget({ force })
+			.catch (err) =>
+				@emitAsync('step-error', err, step)
+				@applyError(err, force)
+
+	applyError: (err, force) =>
+		@_applyingSteps = false
+		@applyInProgress = false
+		@failedUpdates += 1
+		@reportCurrentState(update_failed: true)
+		if @scheduledApply?
+			console.log('Updating failed, but there is already another update scheduled immediately: ', err)
+		else
+			delay = Math.min((2 ** @failedUpdates) * 500, 30000)
+			# If there was an error then schedule another attempt briefly in the future.
+			console.log('Scheduling another update attempt due to failure: ', delay, err)
+			@triggerApplyTarget({ force, delay })
+		@emitAsync('apply-target-state-error', err)
+		@emitAsync('apply-target-state-end', err)
+
+	applyTarget: ({ force = false } = {}) =>
+		console.log('Applying target state')
+		Promise.using @inferStepsLock(), =>
+			Promise.join(
+				@getCurrentForComparison()
+				@getTarget()
+				(currentState, targetState) =>
+					@deviceConfig.getRequiredSteps(currentState, targetState, @stepsInProgress)
+					.then (deviceConfigSteps) =>
+						if !_.isEmpty(deviceConfigSteps)
+							return deviceConfigSteps
+						else
+							@applications.getRequiredSteps(currentState, targetState, @stepsInProgress)
+			)
+			.then (steps) =>
+				if _.isEmpty(steps) and _.isEmpty(@stepsInProgress)
+					console.log('Finished applying target state')
+					@applyInProgress = false
+					@failedUpdates = 0
+					@lastSuccessfulUpdate = Date.now()
+					@reportCurrentState(update_failed: false, update_pending: false, update_downloaded: false)
+					@emitAsync('apply-target-state-end', null)
+					return
+				@reportCurrentState(update_pending: true)
+				Promise.map steps, (step) =>
+					@applyStepAsync(step, { force })
+		.catch (err) =>
+			@applyError(err, force)
+
+	continueApplyTarget: ({ force = false } = {}) =>
+		return if @applyContinueScheduled
+		@applyContinueScheduled = true
+		setTimeout( =>
+			@applyContinueScheduled = false
+			@applyTarget({ force })
+		, 1000)
+		return
+
+	triggerApplyTarget: ({ force = false, delay = 0 } = {}) =>
+		if @applyInProgress
+			if !@scheduledApply?
+				@scheduledApply = { force, delay }
+				@once 'apply-target-state-end', =>
+					@triggerApplyTarget(@scheduledApply)
+					@scheduledApply = null
+			else
+				# If a delay has been set it's because we need to hold off before applying again,
+				# so we need to respect the maximum delay that has been passed
+				@scheduledApply.delay = Math.max(delay, @scheduledApply.delay)
+				@scheduledApply.force or= force
+			return
+		@applyInProgress = true
+		setTimeout( =>
+			@applyTarget({ force })
```

Comment:
Instead of using the event emitter and conditionally adding `.once` handlers you can instead put a `.finally` here and if `@scheduledApply` is set, re-run. This way you avoid having to deal with event handlers above

Reply:
Working on this one... it's a bit more complicated than that because of the way applyTarget works - it doesn't return a promise that gets resolved when everything finishes. Instead several steps will run in parallel and whenever each of them finishes we use continueApplyTarget to schedule another run of applyTarget.

I'm quite sure I can make it work purely with promises but I need to think it through to avoid messing up.